proto/buf/alpha/registry/v1alpha1/schema.proto

```diff
@@ -32,48 +31,54 @@ service SchemaService {
 }
 
 message GetSchemaRequest {
-  // The module from which the schema is retrieved. Required. This refers to a
-  // module that exists in the same BSR instance as is handling this RPC.
-  LocalModuleReference module = 1;
-
-  // One or more element names. The names may be type names (messages or enums),
+  // The owner of the repo that contains the schema to retrieve (a user name or
+  // organization name).
+  string owner = 1;
+  // The name of the repo that contains the schema to retrieve.
+  string repository = 2;
+  // Optional version of the repo. If unspecified, defaults to latest version on
+  // the repo's "main" branch.
+  string reference = 3;
+
+  // Zero or more element names. The names may be type names (messages or enums),
   // service names, or method names. All names must be fully-qualified. If any name
   // is unknown, the request will fail and no schema will be returned.
   //
-  // The resulting schema contains these elements and all of their dependencies,
-  // which is enough information for the caller to construct a dynamic message for
-  // any requested types (e.g. requested message types or request and response types
-  // for requested service and method names).
-  repeated string element_names = 2;
-
-  // If true, well-known imports will be included the returned set of files. If
-  // false or not present, these files will omitted from the response (e.g. the
-  // caller is assumed to already know the types they define).
+  // If no names are provided, the full schema for the module is returned. Otherwise,
+  // the resulting schema contains only the named elements and all of their
+  // dependencies, which is enough information for the caller to construct a dynamic
+  // message for any requested types (e.g. requested message types or request and
+  // response types for requested service and method names).
+  repeated string element_names = 4;
+
+  // If present, this is a commit ID that the client already has cached. So if
+  // the given module reference resolves to this same version, the server should
+  // not send back any descriptors since the client already has them.
   //
-  // These files are a subset of the files in the "well-known types" module:
-  // https://buf.build/beta/wellknowntypes
+  // This allows a client to efficiently poll for updates: after the initial RPC
+  // to get a schema, the client can cache the descriptors and the resolved
+  // version. It then includes that version in subsequent requests in this field,
+  // and the server will only reply with a schema (and new version number) if/when
+  // the resolved version changes.
+  string if_not_version = 5;
```

Comment:
feels like something that should be in a header and  generic

Reply:
_For posterity, here are remarks from a Zoom discussion:_

Indeed! In fact, if this were a simple HTTP GET endpoint, it would be an `If-Not-Modified` header. But an HTTP server has clear semantics for that header and can return a 304 status code to indicate that it decided to not return a response. But for RPC, those semantics don't really apply. This value necessarily impacts the behavior of the operation, causing it to return a nil file descriptor set in the response.

I don't really like putting behavior-changing inputs in metadata. It's more awkward to document since it has to just be prose in the method comment, it's not discoverable in an IDE since auto-completion won't allow a caller to discover it, and it's not type-safe (not a big deal in this case since the type of this field is just a string).

If there were a _standard_ for something like this, I'd be much more amenable to demoting this out of the request. But without such a standard for this kind of behavior, I think this really should be a request field.

In the future, we can wrap this RPC endpoint In a simple HTTP GET API, wherein this will indeed be a header. (And the version in the response message will  be an "Etag" response hedaer.) And when the handler replies with no file descriptor set, the HTTP handler can return a "304 Not Modified".